docs/src/checkpointer.md

```diff
@@ -1,12 +1,136 @@
+```
+
+### Reading checkpoints
+
+There are two ways to restart a simulation from checkpoints. By default,
+`ClimaCoupler` tries finding suitable checkpoints and automatically use them.
+Alternatively, you can specify a directory and a simulation time and restart
+from files saved in the given directory at the given time. If the model you are
+running supports writing checkpoints via command-line argument, it will probably
+also support reading them. In this case, the arguments `restart_dir` and
+`restart_t` identify the path of the top level directory containing all the
+checkpoint files and the simulated times in second.
+
+If the model does not support directly reading a checkpoint, the `Checkpointer`
+module provides a straightforward way to add this feature.
+[`Checkpointer.restart!`](@ref) takes a coupled simulation, a `restart_dir`, and
+a `restart_t` and overwrites the content of the coupled simulation with what is
+in the checkpoint. [`Checkpointer.auto_restart!`](@ref) does the same but trying
+to automatically detect checkpoint files.
+
+## Developer notes
+
+In theory, the state of the component models should fully determine the state of
+the coupled simulation and one should be able to restart a coupled simulation
+just by using the states of the component models. Unfortunately, this is
+currently not the case in `ClimaCoupler`. The main reason for this is the
+complex interdependencies between component models and within `ClimaAtmos` which
+make the initialization step inconsistent. For example, in a coupled simulation,
+the surface albedo should be determined by the surface models and used by the
+atmospheric model for radiation transfer, but `ClimaAtmos` also tries to set the
+surface albedo (since when it has to do so when run in standalone mode). In
+addition to this, `ClimaAtmos` has a large cache that has internal
+interdependencies that are hard to disentangle, and changing a field might
+require changing some other field in a different part of the cache. As a result,
+it is not easy for `ClimaCoupler` to consistently do initialization from a cold
+state. To conclude, restarting a simulation exclusively using the states of the
+component models is currently impossible.
+
+Given that restarting a simulation from the state is impossible, `ClimaCoupler`
+needs to save the states and the caches. Let us review how we use
+`ClimaCore.InputOutput` and `JLD2` package to accomplish this.
+
+`ClimaCore.InputOutput` provides a loss-less way to save the content of certain
+`ClimaCore` objects to HDF5 files. Objects saved in this way are not tied to a
+particular computing device or configuration. When running with MPI,
+`ClimaCore.InputOutput` are also efficiently written in parallel.
+
+Unfortunately, `ClimaCore.InputOutput` only supports certain objects, such as
+`Field`s and `Space`s, but the cache in component models is more complex than
+this and contains complex objects with highly stateful quantities (e.g., C
+pointers).
```

Comment:
After re-reading, I understand that this is why we save the state to HDF5 and cache to JLD2 (I left a comment about this somewhere). It would be good to say that explicitly here.
```suggestion
Unfortunately, `ClimaCore.InputOutput` only supports certain objects, such as
`Field`s and `Space`s, but the cache in component models is more complex than
this and contains complex objects with highly stateful quantities (e.g., C
pointers). Because of this, model states are saved to HDF5 but caches must be
saved to JLD2 files.
```